url3f8d6440dd_huiwq1990_CognativeModel_tgz-pJ8-clarion_samples_HelloWorldJ8/src/clarion/system/MS.java

```diff
@@ -0,0 +1,805 @@
+	private BehavioralInhibitionSystem BIS = new BehavioralInhibitionSystem ();
+	
+	/**The system for drives that belong to neither behavioral system*/
+	private NeitherBehavioralSystem NEITHER = new NeitherBehavioralSystem ();
+	
+	/**The current input (including current sensory information, current goal, chunks in 
+	 * working memory, and drive stimulus).*/
+	private DimensionValueCollection CurrentInput;
+	
+	/**
+	 * Minimally initializes the MS. This constructor builds a "frame" for this instance of
+	 * the MS from which all desired components can be attached. During initialization
+	 * this instance of the MS will attach itself to the CLARION agent you specify.
+	 * @param Agent The agent to which the subsystem is being attached.
+	 */
+	public MS (CLARION Agent) 
+	{
+		super(Agent);
+	}
+	
+	/**
+	 * Gets the drive strength for the drive with the specified ID. If none of the 
+	 * behavioral systems have a drive with the specified ID, this method returns null.
+	 * <p>
+	 * If you are using the drive equation, the specified drive input MUST have a dimension whose ID is 
+	 * equal to the specified ID and that dimension must contain a value with the ID: STIMULUS 
+	 * (of enumerated type RequiredEquationInputs located in the from the DriveEquation class).
+	 * @param ID The ID of the drive whose drive strength you want to get.
+	 * @param DriveInput The current input for the MS.
+	 * @return The strength of the drive with the specified ID.
+	 */
+	public DriveStrength calculateDriveStrength (Object ID, DimensionValueCollection DriveInput)
+	{
+		CurrentInput = DriveInput.clone();
+		Drive d = BAS.get(ID);
+		if(d != null)
+		{
+			Dimension dim = CurrentInput.get(d.getID());
+			if (dim == null)
+			{
+				dim = new Dimension (d.getID());
+				CurrentInput.put(dim.getID(),dim);
+			}
+			Value val = dim.get(Drive.TypicalInputs.SYSTEM_GAIN);
+			if (BIS.containsKey(d.getID()))
+			{
+				if (val == null)
+				{
+					dim.put(Drive.TypicalInputs.SYSTEM_GAIN,
+							new Value(Drive.TypicalInputs.SYSTEM_GAIN, 
+							(BAS.GAIN + BIS.GAIN)/2));
+				}
+				else
+				{
+					val.setActivation((BAS.GAIN + BIS.GAIN)/2);
+				}
+			}
+			else
+			{
+				if (val == null)
+				{
+					dim.put(Drive.TypicalInputs.SYSTEM_GAIN,
+							new Value(Drive.TypicalInputs.SYSTEM_GAIN,BAS.GAIN));
+				}
+				else
+				{
+					val.setActivation(BAS.GAIN);
+				}
+			}
+			return d.calculateDriveStrength(CurrentInput);
+		}
+		d = BIS.get(ID);
+		if(d != null)
+		{
+			Dimension dim = CurrentInput.get(d.getID());
+			if (dim == null)
+			{
+				dim = new Dimension (d.getID());
+				CurrentInput.put(dim.getID(),dim);
+			}
+			Value val = dim.get(Drive.TypicalInputs.SYSTEM_GAIN);
+			if (val == null)
+			{
+				dim.put(Drive.TypicalInputs.SYSTEM_GAIN,
+						new Value(Drive.TypicalInputs.SYSTEM_GAIN,BIS.GAIN));
+			}
+			else
+			{
+				val.setActivation(BIS.GAIN);
+			}
+			return d.calculateDriveStrength(CurrentInput);
+		}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 168:186, 191:209.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.